Allow use of TTree::GetEntry in PoolSource

Dr15Jones opened the pull request:
#### PR description:

When specifying `delayReadingEventProducts == False` in PoolSource this now uses TTree::GetEntry when reading all the data products. This allows ROOT to use implicit multi-threading to decompress multiple TBranches concurrently.


#### PR validation:

This was tested against an actual low efficiency AOD to MiniAOD production job. CPU and number of IO operation measurements show this change to be a substantial improvement over the original implementation of `delayReadingEventProducts == False`.

resolves https://github.com/cms-sw/framework-team/issues/1380

Dr15Jones committed:
Improved RootFile constructor

Grouped related items together.
Removed unnecessary member data.

Dr15Jones committed:
Prompt reading uses TTree::GetEntry

When the source sets prompt reading of Event products, we now use a new delayed reader which makes use of TTree::GetEntry to allow concurrent branch processing.

Dr15Jones committed:
Protect setting EDProductGetter with a guard

Dr15Jones committed:
Handle multiple threads reading RefCores

Deal with the case where the call to TTree::GetEntry uses multiple threads to de-serialize data from an Event.

Dr15Jones committed:
Use just one TTreeCache when prompt reading

Dr15Jones committed:
Added test for Ref use while promptReading

Dr15Jones committed:
Changing TTreeCache now requires specifying the TTree

Dr15Jones committed:
Use separate TTreeCache for auxiliary branch

Dr15Jones committed:
Protect constructing and setting TTreeCache

Manipulating the TTreeCache in ROOT is fairly brittle so need to have extra safe-guards.

makortel committed:
Read event provenance data promptly as well

Dr15Jones committed:
Removed extra call to SetEnablePrefetching

Dr15Jones committed:
Code formatting

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48088/44832

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones for master.

It involves the following packages:

- DataFormats/Common (**core**)
- DataFormats/FWLite (**core**)
- FWCore/Framework (**core**)
- FWCore/Integration (**core**)
- IOPool/Input (**core**)
- IOPool/Output (**core**)
- IOPool/Streamer (**core**)


@Dr15Jones, @cmsbuild, @makortel, @smuzaffar can you please review it and eventually sign? Thanks.
@makortel, @missirol, @mmusich, @rovere, @wddgit this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
+1

**Size**: This PR adds an extra 424KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2607da/46163/summary.html
**COMMIT**: d603932b8bd073af7cc5934dd380f4d3a8bb794e
**CMSSW**: CMSSW_15_1_X_2025-05-15-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/48088/46163/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-15-1100+2607da/68513/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 2 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-15-1100+2607da/68513/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4038193
 - DQMHistoTests: [Total failures: 0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-15-1100+2607da/68513/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4038173
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-15-1100+2607da/68513/triggerResults): no differences found

makortel commented on IOPool/Input/src/RootTree.cc:
```diff
@@ -549,18 +605,13 @@ namespace edm {
-      tree->SetCacheSize(cacheSize);
-      std::unique_ptr<TTreeCache> treeCache(dynamic_cast<TTreeCache*>(file.GetCacheRead()));
+      std::unique_ptr<TTreeCache> treeCache = file.createCacheWithSize(*tree, cacheSize);
       if (nullptr != treeCache.get()) {
         treeCache->StartLearningPhase();
         treeCache->SetEntryRange(0, tree->GetEntries());
         treeCache->AddBranch(branchNames, kTRUE);
         treeCache->StopLearningPhase();
       }
```
I'm wondering the the `treeCache` should be active in the `TFile` during these `treeCache` calls (also elsewhere).

makortel replied:
According to @pcanal `treeCache` does not need to be active in the `TFile` during these calls.

makortel commented:
FYI @dan131riley

Dr15Jones commented on IOPool/Input/src/InputFile.h:
```diff
@@ -46,8 +48,30 @@ namespace edm {
+      TObject* tree_;
+    };
+    [[nodiscard]] CacheGuard setCacheReadTemporarily(TFileCacheRead* tfcr, TObject* iTree) {
+      file_->SetCacheRead(tfcr, iTree, TFile::kDoNotDisconnect);
+      return CacheGuard(file_.get(), iTree);
```
I guess I'm indifferent to the idea.

Dr15Jones replied:
made the change

Dr15Jones commented on IOPool/Input/src/RootFile.cc:
```diff
@@ -1372,6 +1347,7 @@ namespace edm {
       }
       eventHistoryBranch->SetAddress(&pHistory);
       roottree::getEntry(eventHistoryTree_, eventTree_.entryNumber());
+      eventHistoryBranch->SetAddress(nullptr);
```
paranoia?

pcanal replied:
Since `pHistory` is a local variable and thus has a shorter lifetime that the `TTree` (`eventHistoryTree_`). This is semantically the right choice.

Dr15Jones committed:
Improvements based on review feedback

- explicitly delete compiler generated routines
- improved RAII
- avoid unnecessary temporary
- add arena isolations

Dr15Jones commented on DataFormats/Common/interface/RefCoreStreamer.h:
```diff
@@ -33,7 +33,25 @@ namespace edm {
+  class RefCoreStreamerGuard {
+  public:
+    RefCoreStreamerGuard(EDProductGetter const* ep) { setRefCoreStreamer(ep); }
+    ~RefCoreStreamerGuard() { unsetRefCoreStreamer(); }
```
done

Dr15Jones commented on IOPool/Input/src/RootPromptReadDelayedReader.h:
```diff
@@ -0,0 +1,82 @@
+    edm::propagate_const<std::shared_ptr<InputFile>> filePtr_;
+    edm::propagate_const<DelayedReader*> nextReader_;
+    std::unique_ptr<SharedResourcesAcquirer>
+        resourceAcquirer_;  // We do not use propagate_const because the acquirer is itself mutable.
```
done

Dr15Jones commented on IOPool/Input/src/RootTree.cc:
```diff
@@ -153,14 +166,29 @@ namespace edm {
+    assert(nullptr != classCache_);
+    void* p = classCache_->New();
+    std::unique_ptr<WrapperBase> edp = getWrapperBasePtr(p, offsetToWrapperBase_);
+    return edp;
```
done

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48088/44859

Dr15Jones commented on IOPool/Input/src/InputFile.h:
```diff
@@ -46,8 +48,30 @@ namespace edm {
+      CacheGuard() = delete;
+      CacheGuard(CacheGuard const&) = delete;
+      CacheGuard& operator=(CacheGuard const&) = delete;
+      ~CacheGuard() { file_->SetCacheRead(nullptr, tree_, TFile::kDoNotDisconnect); }
```
done

Dr15Jones commented:
please test

cmsbuild commented:
Pull request #48088 was updated. @Dr15Jones, @makortel, @smuzaffar can you please check and sign again.

cmsbuild commented:
+1

**Size**: This PR adds an extra 36KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2607da/46223/summary.html
**COMMIT**: 926a864e8fdfc945587e8e95ced5e273e14ab706
**CMSSW**: CMSSW_15_1_X_2025-05-16-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/48088/46223/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-16-1100+2607da/68560/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 6 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-16-1100+2607da/68560/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4038193
 - DQMHistoTests: [Total failures: 22](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-16-1100+2607da/68560/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4038151
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-16-1100+2607da/68560/triggerResults): no differences found

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48088/44871

cmsbuild commented:
+1

**Size**: This PR adds an extra 24KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2607da/46244/summary.html
**COMMIT**: 7618354323393749d1c73c16a4d88faabfdde885
**CMSSW**: CMSSW_15_1_X_2025-05-19-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48088/46244/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-19-1100+2607da/68589/validateJR/logRootQA-events.log) from the logs
 - [Reco comparison results: 2 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-19-1100+2607da/68589/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4038193
 - DQMHistoTests: [Total failures: 25](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-19-1100+2607da/68589/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4038148
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-19-1100+2607da/68589/triggerResults): no differences found

makortel commented:
We measured separately that this PR would use ~100 MB more memory in the re-mini-nano jobs that would use prompt reading (compared to present prompt reading). This isn't visible in the MaxMemoryPreload report though, because none of the tested runTheMatrix workflows seem to run the re-mini as the first step. (there are variations likely related to https://github.com/cms-sw/cmssw/issues/46966)

cmsbuild commented:
+1

**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-2607da/46278/summary.html
**COMMIT**: 7618354323393749d1c73c16a4d88faabfdde885
**CMSSW**: CMSSW_15_1_X_2025-05-20-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48088/46278/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 1 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-20-2300+2607da/68611/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 9 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-20-2300+2607da/68611/validateJR.html)
 - DQMHistoTests: Total files compared: 58
 - DQMHistoTests: Total histograms compared: 4098225
 - DQMHistoTests: [Total failures: 12](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-20-2300+2607da/68611/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4098193
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 57 files compared)
 - Checked 243 log files, 200 edm output root files, 58 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-20-2300+2607da/68611/triggerResults): no differences found

makortel commented:
2500.021
<img width="703" alt="image" src="https://github.com/user-attachments/assets/3f02bfd8-da7c-47d8-9f9c-a041be376e1e" />

2500.022
<img width="703" alt="image" src="https://github.com/user-attachments/assets/9cfe6079-fe51-4839-8fdb-9b09a00c55d6" />


2500.023
<img width="703" alt="image" src="https://github.com/user-attachments/assets/e212f3fb-fa11-43b3-bad4-7391f1e74fb9" />

2500.024
<img width="703" alt="image" src="https://github.com/user-attachments/assets/8582a5e4-7911-458b-ba65-1c724c7f40b7" />

2500.031
<img width="703" alt="image" src="https://github.com/user-attachments/assets/5267240d-26af-4bfa-950c-52ff826d7f37" />

2500.032
<img width="703" alt="image" src="https://github.com/user-attachments/assets/b85b53a8-96f2-4c3b-9b05-d47108edf966" />

2500.033
<img width="703" alt="image" src="https://github.com/user-attachments/assets/2e310244-455e-4082-8f14-981987b11f28" />

2500.034
<img width="703" alt="image" src="https://github.com/user-attachments/assets/3eae38f4-541e-49e2-a7ae-b40014457a07" />

makortel commented:
To me these numbers are consistent with the ~100 MB increase expectation convolved with the variations mentioned in https://github.com/cms-sw/cmssw/issues/46966

makortel commented:
+core

mandrenguyen commented:
+1
